### PR TITLE
aiesim support and more tutorial updates.

### PR DESCRIPTION
**Pull request by jackl-xilinx**
Add preliminary support for aiesim. By default, we auto-generate the Work folder which generates some files and copies in others. Please review implementation and comment if there's a better way to integrate the Work folder generation.

### CODE REVIEW

File: runtime_lib/test_library.h [hunk 1]
```diff
@@ -42,6 +57,11 @@ extern "C" {
 #define MODE_PL 1
 #define MODE_MEM 2
 
+//#define HIGH_ADDR(addr)	((addr & 0xffffffff00000000) >> 32)
+//#define LOW_ADDR(addr)	(addr & 0x00000000ffffffff)
+
+//#define MLIR_STACK_OFFSET 4096
+
```
Comment (fifield) [inline runtime_lib/test_library.h:64]: Remove this

**Comment by jackl-xilinx** [comment]
> I look forward to trying this out. I had a few minor comments as I looked through.
> 
> Are there any tests for this or only the tutorial examples?

Thanks Jeff for all the great feedback. I'm very open to changing how we generate the Work folder files to make it more streamlined and extensible. There are no tests for this that's checked in yet. I just wanted to get the feature in first. But you're right, we should check in some built-in tests with the feature to really ensure it's working when built.

**Comment by jamesroxbypb** [comment]
If we are looking to build up more "board" tests - ie. tests to run on an actual board as part of the VCK5000 and eventual V70 bringup which report PASS/FAIL rather than being examples, it might be that we would write them in such a way that we could test both physical hardware and the simulator at the same time.

**Review by denolf — APPROVED** [review]
Great new feature.

File: tools/aiecc/aiecc/main.py [hunk 2]
```diff
@@ -265,6 +265,38 @@ async def process_arm_cgen(self):
...
+        os.mkdir('Work/arch')
+        os.mkdir('Work/reports')
+        os.mkdir('Work/config')
+        os.mkdir('Work/ps')
+        os.mkdir('Work/ps/c_rts')
+        os.mkdir('Work/ps/c_rts/systemC')
+        os.mkdir('Work/ps/c_rts/systemC/generated-source')
+        os.mkdir('Work/ps/c_rts/systemC/generated-objects')
+      except FileExistsError:
+        pass
```
[the remark this answers is not in the document]
Comment (jackl-xilinx) [reply tools/aiecc/aiecc/main.py]: Thanks. I've changed it to use makdirs with exist_ok=True so it should continue even if it exists.

File: tools/aiecc/aiecc/main.py [hunk 3]
```diff
@@ -328,6 +360,8 @@ async def run_flow(self):
         progress.task_completed = progress.add_task("[green] AIE Compilation:", total=len(cores)+1, command="%d Workers" % nworkers)
 
         processes = [self.process_arm_cgen()]
+        if(opts.aiesim):
```
[the remark this answers is not in the document]
Comment (jackl-xilinx) [reply tools/aiecc/aiecc/main.py:358]: I'm not sure I understand why this check is needed if we're always running libxaiev2, or maybe it's compared to vck5000 case. In either case, I wasn't sure how to add this check so I may add it later.

File: runtime_lib/aiesim/Makefile [hunk 4]
```diff
@@ -0,0 +1,18 @@
+ifeq ($(AIE_SRC_DIR),)
```
[the remark this answers is not in the document]
Comment (jackl-xilinx) [reply runtime_lib/aiesim/Makefile:6]: Done.

File: lib/Targets/AIETargetSimulationFiles.cpp [hunk 5]
```diff
@@ -0,0 +1,194 @@
+//===- AIETargetSimulationFiles.cpp -----------------------------*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2021 Xilinx Inc.
```
[the remark this answers is not in the document]
Comment (jackl-xilinx) [reply lib/Targets/AIETargetSimulationFiles.cpp]: Done.

File: lib/Targets/AIETargetSimulationFiles.cpp [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+  /* Generates a aieshim_solution.aiesol file which is necessary to run aiesim.
+  Sample invocation:
+  aie-translate --aie-mlir-to-shim-solution ./aie.mlir >
+  ./Work/arch/aieshim_solution.aiesol
+
+  NOTE: to correctly enable all tiles used for routing, the aie-opt routing pass
+  must be called first. So, a more practical invocation: aie-opt
+  --aie-create-pathfinder-flows ./aie.mlir | aie-translate --aie-mlir-to-shim >
+  ./Work/arch/aieshim_solution.aiesol
+  */
```
[the remark this answers is not in the document]
Comment (jackl-xilinx) [reply lib/Targets/AIETargetSimulationFiles.cpp]: Done.

File: runtime_lib/test_library.h (hunk 1 above)
Comment (jackl-xilinx) [reply runtime_lib/test_library.h]: Done.

**Comment by AndraBisca** [comment]
Thank you for all the work! I am looking forward to trying it out!

**Comment by jackl-xilinx** [comment]
Simulation folder renamed from Work to sim and structure is simplified. Now, you don't need your own Makefile. You can call make -C sim; make -C sim sim to compile and run aiesimulator directly. It still has hooks to test.cpp which needs a custom generator for genwrapper_for_ps.cpp to fix but at the moment, we're just using a template file. We can push this version and then somebody else can add that feature support to make it more robust.

File: tools/aiecc/aiecc/main.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (fifield) [reply tools/aiecc/aiecc/main.py:358]: Then maybe the correct thing is to remove the v1 code from test_library.cpp code in a different PR.

**Comment by fifield** [comment]
> If we are looking to build up more "board" tests - ie. tests to run on an actual board as part of the VCK5000 and eventual V70 bringup which report PASS/FAIL rather than being examples, it might be that we would write them in such a way that we could test both physical hardware and the simulator at the same time.

This is the "simulator is just a different board" approach which is great because it means we can actually run tests without a board. In the meantime it would be good to have at least one test enabled to run the simulation flow end-to-end so that we know when something breaks.

**Pull request merged**